jaeger-core/src/main/java/io/jaegertracing/senders/SenderResolver.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2018, The Jaeger Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.jaegertracing.senders;
+
+import io.jaegertracing.Configuration;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Provides a way to resolve an appropriate {@link Sender}
+ */
+@Slf4j
+public class SenderResolver {
+
+  /**
+   * Resolves a {@link Sender} by passing {@link Configuration.SenderConfiguration#fromEnv()} down to the
+   * {@link SenderFactory}
+   *
+   * @see #resolve(Configuration.SenderConfiguration)
+   * @return the resolved Sender, or NoopSender
+   */
+  public static Sender resolve() {
+    return resolve(Configuration.SenderConfiguration.fromEnv());
+  }
+
+  /**
+   * Resolves a sender by passing the given {@link Configuration.SenderConfiguration} down to the
+   * {@link SenderFactory}. The factory is loaded either based on the value from the environment variable
+   * JAEGER_SENDER_FACTORY or, in its absence or failure to deliver a {@link Sender}, via the {@link ServiceLoader}.
+   * If no factories are found, a {@link NoopSender} is returned. If multiple factories are available, the
+   * factory whose {@link SenderFactory#getType()} matches the JAEGER_SENDER_FACTORY env var is selected.
+   * If none matches, {@link NoopSender} is returned.
+   *
+   * @param senderConfiguration the configuration to pass down to the factory
+   * @return the resolved Sender, or NoopSender
+   */
+  public static Sender resolve(Configuration.SenderConfiguration senderConfiguration) {
+    String senderFactoryClass = System.getProperty(Configuration.JAEGER_SENDER_FACTORY);
```

Comment:
nit: `senderFactoryClass` -> `senderFactoryType` ?

Reply:
Isn't this still using the old mechanism - where the env var specifies a fq class name?

Reply:
It can be both. At this point in the code, we assume it's a class (we try to use it on a `Class.forName` later on).

Reply:
> Isn't this still using the old mechanism - where the env var specifies a fq class name?

Yes. I thought the type is only used to disambiguate in case there are multiple factories available? The previous resolution was:

- FQCN
- Service Loader (first service available)
- NoopSender

The current is:

- FQCN
- Service Loader (when only one service is available)
- Service Loader + Type (when multiple services are available)
- NoopSender

Reply:
Type in `SenderFactory` seems redundant as you can still specify the wanted impl as FQCN or wouldn't it work?

Reply:
It would, and it's how it was being done before. Looks at this thread: https://github.com/jaegertracing/jaeger-client-java/pull/449#discussion_r195080992